### ISSUE TITLE
`-shared` flag to emcc causes standalone wasm build for .js target

### ISSUE THREAD

**Issue by hostilefork** [emscripten-core/emscripten#11814]
I'm not entirely sure why the build command line we were using has `-shared` on it, but it did... :-/

One consequence of this is that when using `tot-upstream` with recent changes, if you say:

     $ emcc -shared test.c -o test.js

It will produce a test.js file that is actually Wasm data (and no accompanying .wasm).  So presumably it is cuing a standalone build without warning the extension doesn't match.

Removing the `-shared` makes it act as before...producing both a .js and .wasm file.

Reproduces on both Linux and Mac.  Again, that's with `tot-upstream`.  (emcc reports (Emscripten gcc/clang-like replacement) 1.40.1 (68f3899f2e3570ca8e7866853ddc90f175406349))

**Comment by sbc100** [comment]
What are you expecting the output of `-shared` to be exactly?   Are you planning on loading that output directly?   

It sounds like to you are expecting the result of the build to be exactly the same as if you hasn't passed `-shared`?  I guess before my recent change we were simply ignoring this last and so you were getting an executable.

Traditionally in emscripten if you build a "shared libraries" or `.so` file that is expected to be passed into emcc again for the final link.    

We don't have a concept in emscripten of a shared library that can be run directly.

**Comment by sbc100** [comment]
I guess I can restore the pevious behaviour of ignore the `-shared` flag if the output extension is JS, at least for some period of time.

Could you investigate why your build system is passing `-shared` to something your are then directly using?  One solution is run a final link to turn your shared library into an executable:

```
emcc -shared -o mylib.so      #  build a shared library
emcc -o app.js mylib.so         # produce an application that links in the shared library statically
```

**Comment by hostilefork** [comment]
> Could you investigate why your build system is passing -shared to something your are then directly using?

As opposed to thinking of it as "being run directly", it's a matter of sending the process down the path of building something that does not have a main().

The other build targets make executables, but this makes a library that has to be initiated from JavaScript.  So it is more "DLL-like" in that way...thus going down the build switches that apply for that.  Sticking -shared on the command line is something the other non-main()-having targets do.

> I guess I can restore the pevious behaviour of ignore the -shared flag if the output extension is JS, at least for some period of time.

It's not an issue if it is kept this way.  But seems it would help a lot for people in the future in finding out what went wrong if there were a warning when .js extension products were built as wasm!  (I didn't know what was going on and spent a while thinking there was a problem with the uploading process to s3...)

**Comment by sbc100** [comment]
I see.  Thank you for the explanation.

In emscripten you don't need to used `-shared` to produce and application without a main.   There are other ways to do that (we should probably document them better).

I will restore the previous behaviour in the case the that `-shared` is used with the `.js` output extension.    However I will add a warning to suggest simply dropping the `-shared` flag.

**Comment by sbc100** [comment]
This was fixed in #11820
